modules/security-best-practices/pages/rust-canister-development-security-best-practices.adoc

```diff
@@ -0,0 +1,414 @@
+
+=== Consider using stable memory, version it, test it
+
+==== Security Concern
+
+Canister memory is not persisted across upgrades. If data needs to be kept across upgrades, a natural thing to do is to serialize the canister memory in `pre_upgrade`, and deserialize it in `post_upgrade`. However, the available number of instructions for these methods is limited. If the memory grows too big, the canister can no longer be updated.
+
+==== Recommendation
+
+- Stable memory is persisted across upgrades and can be used to address this issue.
+- link:https://mmapped.blog/posts/01-effective-rust-canisters.html#stable-memory-main[Consider using stable memory.] (from link:https://mmapped.blog/posts/01-effective-rust-canisters.html[Effective Rust Canisters]). See also the disadvantages discussed there.
+- link:https://mmapped.blog/posts/01-effective-rust-canisters.html#version-stable-memory[Version stable memory.] (from link:https://mmapped.blog/posts/01-effective-rust-canisters.html[Effective Rust Canisters])
+- link:https://mmapped.blog/posts/01-effective-rust-canisters.html#test-upgrades[Test the upgrade hooks.] (from link:https://mmapped.blog/posts/01-effective-rust-canisters.html[Effective Rust Canisters])
+- See also the section on upgrades in link:https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister[How to audit an Internet Computer canister] (though focused on Motoko)
+- Write tests for stable memory to avoid bugs.
+- Some libraries (mostly work in progress / partly unfinished) that people work on:
+** link:https://github.com/dfinity/stable-structures/[https://github.com/dfinity/stable-structures/]
+*** HashMap: link:https://github.com/dfinity/stable-structures/pull/1[https://github.com/dfinity/stable-structures/pull/1] (currently not prod ready)
+** link:https://github.com/seniorjoinu/ic-stable-memory-allocator[https://github.com/seniorjoinu/ic-stable-memory-allocator]
+- See link:https://wiki.internetcomputer.org/wiki/Current_limitations_of_the_Internet_Computer[Current limitations of the Internet Computer], sections "Long running upgrades" and "[de]serialiser requiring additional wasm memory"
+- For example, link:https://github.com/dfinity/internet-identity[internet identity] uses stable memory directly to store user data.
+
+=== Consider encrypting sensitive data on canisters
+
+==== Security Concern
+
+By default, canisters provide integrity but not confidentiality. Data stored on canisters can be read by nodes / replicas.
+
+==== Recommendation
+
+- Consider end-to-end encrypting any private or personal data (e.g. user’s personal or private information) on canisters.
+- We are currently working on an example dApp (encrypted notes) to illustrate how end-to-end encryption can be done.
+
+=== Create backups
+
+==== Security Concern
+
+Bugs in the upgrade methods or inconsistent state could render a canister unusable.
```

Comment:
Not sure it is quite clear whose bugs this is. Some people may think this refers to bugs in the IC, I would rewrite as: 

"If a canister has a faulty upgrade process (due to some bug from the dapp developer), then the canister can never be upgraded again. Make sure methods used in upgrading are tested or the canister becomes immutable."

Reply:
I agree this is unclear. Thanks for pointing this out. I changed it along the lines of your suggestion, hope it is clear now!